File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.js
/**
 * @param {number[][]} points
 * @param {number} k
 * @return {number[][]}
 */
var kClosest = function(points, k) {
    const dist = points.map(point => {
        let [x, y] = point;
        return [point, x * x + y * y];
    })

    dist.sort((a, b) => a[1] - b[1]);
    return dist.slice(0, k).map(a => a[0]);
};

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> 
            (b[0] * b[0] + b[1] * b[1]) - (a[0] * a[0] + a[1] * a[1])
        );
        
        for (int[] point : points) {
            pq.add(point);
            if (pq.size() > k) {
                pq.poll(); 
            }
        }

        int[][] res = new int[k][2];
        int index = 0;

        while (!pq.isEmpty()) {
            res[index++] = pq.poll();
        }
        
        return res;
    }
}